### PR TITLE
 Issue-34 Table Attribute Testing - Convert Special Characters in Column Names

### PR DESCRIPTION
**Pull request by Netravati1205**
Added Test case to test the Table attribute - Convert Special Characters in Column Names.

Description - Basically this test case will perform CRUD operation on the DB table where the table attribute Convert Special Characters in Column Names is set to YES.
The table is already re-compiled, when the attribute is set the special character in the column name will be replaced as "_"  and still functionality work as expected

image

Created following object types.
**Fields/$$@#.fld.yml
Fields/@#$.fld.yml
Fields/@@@@@@@@@.fld.yml
Files/@@#$.DEVWBPLIBF.file.yml
Functions/VI0034A.VTI0034.fun.yml
Functions/VI0034C.VTI0034.fun.yml
Lists/LI0034.lsts.yml
Processes/VTI0034.pro.yml**

**Verifier Test report - Result**

===================================
CCS: I0034
Partition: WBP
Language: ENG 
Function/Process: VI0034A / VTI0034
Environment Type: WIN32 32-bit
Environment Path: 'C:\PROGRA~2\Lansa\X_WIN95\X_LANSA\'
DateTime: 2024-01-14 13:39:14
Database: MSSQLS / LANSA
-----
+ Windows.
    2024-01-14 13:39:14 + Test Function: 1, 0 skipped
    2024-01-14 13:39:14  (1) - VI0034C
    2024-01-14 13:39:14 + (1) - Calling Function: VI0034C
    2024-01-14 13:39:14 - VI0034C. Start - Convert Special Characters
    2024-01-14 13:39:14 -  Clear Data
    2024-01-14 13:39:14 -  - 5 record(s) found
    2024-01-14 13:39:14 -  - Successful
    2024-01-14 13:39:14 -  Insert: 5 record(s)
    2024-01-14 13:39:14 -  - Successful
    2024-01-14 13:39:14 -  Update: 5 record(s)
    2024-01-14 13:39:14 -  - Successful
    2024-01-14 13:39:14 -  Test the updated : 5 record(s)
    2024-01-14 13:39:14 -  - Successful
    2024-01-14 13:39:14 - VI0034C. End
    2024-01-14 13:39:14 Completed with <OK>.
===================================
CCS: I0034
Partition: WBP
Language: ENG 
Function/Process: VI0034A / VTI0034
Environment Type: WIN32 32-bit
Environment Path: 'C:\PROGRA~2\Lansa\X_WIN95\X_LANSA\'
DateTime: 2024-01-14 13:40:44
Database: MSSQLS / LANSA
-----
+ Windows.
    2024-01-14 13:40:44 + Test Function: 1, 0 skipped
    2024-01-14 13:40:44  (1) - VI0034C
    2024-01-14 13:40:44 + (1) - Calling Function: VI0034C
    2024-01-14 13:40:44 - VI0034C. Start - Convert Special Characters
    2024-01-14 13:40:44 -  Clear Data
    2024-01-14 13:40:44 -  - 5 record(s) found
    2024-01-14 13:40:44 -  - Successful
    2024-01-14 13:40:44 -  Insert: 5 record(s)
    2024-01-14 13:40:44 -  - Successful
    2024-01-14 13:40:44 -  Update: 5 record(s)
    2024-01-14 13:40:44 -  - Successful
    2024-01-14 13:40:44 -  Test the updated : 5 record(s)
    2024-01-14 13:40:44 -  - Successful
    2024-01-14 13:40:44 - VI0034C. End
    2024-01-14 13:40:44 Completed with <OK>.
    2024-01-14 13:40:44 -----

### CODE REVIEW

**Comment by robe070** [comment]
You have not checked in the table. Therefore I have not been able to review your code any further. Obviously it needs to be added to the User List too.

**Comment by Netravati1205** [comment]
Okay , I got that my mistake **@@#$.DEVWBPLIBF.file.yml** table is not checked in. I will check in that and make sure its added to User List too.

**Comment by Netravati1205** [comment]
Hi @robe070 ,
**Files/@@#$.DEVWBPLIBF.file.yml** Table file was missed while commit the changes by mistake, and already this table file is as part of User list Which I have checked in previously.
Thanks

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    Define_Map For(*BOTH) Class(#BIFA002A) Name(#this_rc)
+    
+    #Com_Owner.mthAdd_SubMsg( OK ('Test the updated : ' + #LF_Original_Count.AsString + ' record(s)') '' )
+    
+    #LF_@#$ := 911119
+    Select Fields(*ALL) From_File(@@#$)
+    
+    * Validate all the rows and only #$$@# field has been updated using mthUpdate method
+    
+    If (#LF_@#$ *NE #@#$)
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: This should only be tested if OK.
And all Fields should be tested.

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml [hunk 2]
```diff
@@ -0,0 +1,256 @@
+﻿Function:
+  Enabled For RDMLX:  Yes
+
+  Source: |
+    Function Options(*DIRECT) Rcv_List(#wLKData #wLRet)
+    
+    * ******************************************************************************************************
+    * Test Case Description
+    * Issue #34 Test Convert Special Characters in Column Names
+    * The Test Data is setup with the values that all the Fields will have names in secial charactors
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: Spelling

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    * And the table attributes of Convert Special Characters in Column Names is set to YES,
+    * table has been recompiled, to have the Convert Special Characters in Column Names to be working.
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: This line is superfluous. All tables are compiled in order for them to work.
A comment you put in somewhere had a summary of the functionality. Included the word "CURD", but in Australia we call it "CRUD". Use some or all of that text.

**Review by robe070 — CHANGES_REQUESTED** [review]

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml (hunk 1 above)
Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: Will resolve it as per comment

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml (hunk 2 above)
Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: will correct and commit the updated one

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml (hunk 3 above)
Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: You mean I need replace "table has been recompiled" to "All tables are compiled in order for them to work" ?

Comment (robe070) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: > You mean I need replace "table has been recompiled" to "All tables are compiled in order for them to work" ?

No. Superfluous means not needed.

A comment you put in somewhere had a summary of the functionality. Included the word "CURD", but in Australia we call it "CRUD". Use some or all of that text.

**Comment by robe070** [comment]
The test case fails to run using SuperServer IO. Its a defect in LANSA. So you need to skip executing this test case if its a SuperServer IO test.
I have written a method to extract the various execution parameters. The code is below. Append this to your Function :
```
*
* Local Function
*    #LF_Execution_Type= W, #LF_Local_Platform W, #LF_Remote_Platform=
* SuperServer Function Call
*    #LF_Execution_Type= F, #LF_Local_Platform W, #LF_Remote_Platform= W
*    #LF_Execution_Type= F, #LF_Local_Platform W, #LF_Remote_Platform= I
* SuperServer IO Call
*    #LF_Execution_Type= Q, #LF_Local_Platform W, #LF_Remote_Platform= W
*    #LF_Execution_Type= Q, #LF_Local_Platform W, #LF_Remote_Platform= I
*
Mthroutine Name(mthGetExecutionType)
Define Field(#LF_Execution_Type) Reffld(#Std_Alpha) /* W = Windows, I = IBMi, F = SuperServer Function Call, Q = SuperServer IO */
Define Field(#LF_Local_Platform) Reffld(#BIFA001A) /* W = Windows, I = IBMi */
Define Field(#LF_Remote_Platform) Reffld(#BIFA001A) /* Blank is not remote, W = Windows, I = IBMi */

Get_Entry Number(1) From_List(#wLKData)
#Com_Owner.mthAdd_Msg( OK ('VTF_KEY1= ' + #VTF_KEY1 + ', VTF_KEY2= ' + #VTF_KEY2 + ', VTF_KEY3= ' + #VTF_KEY3) '' )

#LF_Execution_Type := #VTF_KEY1
#LF_Local_Platform := #VTF_KEY1.Substring( 4 1 )
#LF_Remote_Platform := #VTF_KEY1.Substring( 9 1 )

If (#LF_Local_Platform *EQ '')
#LF_Local_Platform := #LF_Execution_Type
Endif
#Com_Owner.mthAdd_Msg( OK ('#LF_Execution_Type= ' + #LF_Execution_Type + ', #LF_Local_Platform ' + #LF_Local_Platform + ', #LF_Remote_Platform= ' + #LF_Remote_Platform) '' )
Endroutine
```

Add a call to this routine after start of testing. That is after this line of code
```
clr_List Named(#wLRet)
#Com_Owner.mthAdd_Msg( OK *Function ('Start - ' + *Function_Desc) )
```

```
#Com_Owner.mthGetExecutionType()
```

Then wrap the entire test code in a test of the execution type like this:
image

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml (hunk 1 above)
Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: Tested based on the 'OK' flag status

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml (hunk 2 above)
Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: Spelling are corrected.

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml (hunk 3 above)
Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: Updated the comment section text as "CRUD".
Removed not needed comment in the code.

**Comment by Netravati1205** [comment]
**mthGetExecutionType** method has been appended to function to skip the execution of this test case if its a SuperServer IO.

Tested by enabling the **LANSA01_P50PGMLIB** from verifier_connection.dat file.

Here is the result.

===================================
CCS: I0034
Partition: WBP
Language: ENG 
Function/Process: VI0034A / VTI0034
Environment Type: WIN32 32-bit
Environment Path: 'C:\PROGRA~2\Lansa\X_WIN95\X_LANSA\'
DateTime: 2024-01-17 02:13:29
Database: MSSQLS / LANSA
-----
+ Windows.
    2024-01-17 02:13:29 + Test Function: 1, 0 skipped
    2024-01-17 02:13:29  (1) - VI0034C
    2024-01-17 02:13:29 + (1) - Calling Function: VI0034C
    2024-01-17 02:13:29 - VI0034C. Start - Convert Special Characters
    2024-01-17 02:13:29 - VTF_KEY1= W, VTF_KEY2= , VTF_KEY3=
    2024-01-17 02:13:29 - #LF_Execution_Type= W, #LF_Local_Platform W, #LF_Remote_Platform=
    2024-01-17 02:13:29 -  Clear Data
    2024-01-17 02:13:29 -  - 5 record(s) found
    2024-01-17 02:13:29 -  - Successful
    2024-01-17 02:13:29 -  Insert: 5 record(s)
    2024-01-17 02:13:29 -  - Successful
    2024-01-17 02:13:29 -  Update: 5 record(s)
    2024-01-17 02:13:29 - VI0034C. End
    2024-01-17 02:13:29 Completed with <OK>.
+ Windows -> iSeries: LANSA01_P50PGMLIB (SuperServer Function Call).
    2024-01-17 02:13:29 + Test Function: 1, 0 skipped
    2024-01-17 02:13:29  (1) - VI0034C
    2024-01-17 02:13:29 + (1) - Calling SuperServer Function: VI0034C
    2024-01-17 02:13:29 - VI0034C. Start - Convert Special Characters
    2024-01-17 02:13:29 - VTF_KEY1= F: W -> I, VTF_KEY2= , VTF_KEY3=
    2024-01-17 02:13:29 - #LF_Execution_Type= F, #LF_Local_Platform W, #LF_Remote_Platform= I
    2024-01-17 02:13:29 -  Clear Data
    2024-01-17 02:13:29 -  - 4 record(s) found
    2024-01-17 02:13:29 -  - Successful
    2024-01-17 02:13:29 -  Insert: 5 record(s)
    2024-01-17 02:13:29 -  - Successful
    2024-01-17 02:13:29 -  Update: 5 record(s)
    2024-01-17 02:13:29 -  - Successful
    2024-01-17 02:13:29 -  Test the updated : 5 record(s)
    2024-01-17 02:13:29 -  - Successful
    2024-01-17 02:13:29 - VI0034C. End
    2024-01-17 02:13:29 Completed with <OK>.
+ Windows -> iSeries: LANSA01_P50PGMLIB (SuperServer I/O).
  -----
  - Environment Type: IBMI 64-bit
  - Environment Path: '/LANSA_p50pgmlib/x_lansa/'
  - DateTime: 2024-01-17 13:13:31
  - Database: DB2ISERIES / *LOCAL
  -----
    2024-01-17 02:13:29 + Test Function: 1, 0 skipped
    2024-01-17 02:13:29  (1) - VI0034C
    2024-01-17 02:13:29 + (1) - Calling Function: VI0034C
    2024-01-17 02:13:29 - VI0034C. Start - Convert Special Characters
    2024-01-17 02:13:29 - VTF_KEY1= Q: W -> I, VTF_KEY2= , VTF_KEY3=
    2024-01-17 02:13:29 - #LF_Execution_Type= Q, #LF_Local_Platform W, #LF_Remote_Platform= I
    2024-01-17 02:13:29 - Skipping SuperServer IO test due to defect described in #34
    2024-01-17 02:13:29 - VI0034C. End
    2024-01-17 02:13:29 Completed with <OK>.
+ Windows -> iSeries: (Evaluation).
    2024-01-17 02:13:29 Skipped.
    2024-01-17 02:13:29 -----

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml [hunk 4]
```diff
@@ -36,11 +35,16 @@
     Clr_List Named(#wLRet)
     #Com_Owner.mthAdd_Msg( OK *Function ('Start - ' + *Function_Desc) )
     
+    #Com_Owner.mthGetExecutionType()
+    
     Get_Entry Number(1) From_List(#wLKData)
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: Delete this line. Its included in mthGetExecutionType()

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml [hunk 5]
```diff
@@ -36,11 +35,16 @@
...
     
     *
     * Add your test case here
     * It should be a list of method calls so its easy to read whats being tested.
+    If (#LF_Execution_Type *EQ Q)
+    #Com_Owner.mthAdd_Msg( OK ('Skipping SuperServer IO test due to defect described in #34') ('') )
```
Comment (robe070) [inline Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: Use the ticket number I gave you Issue #34 does not describe this defect.

Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: As ticket points mentioned that IBM-i are not to be tested Thought to mention #34.

Now its corrected to CCS ticket number

**Review by robe070 — CHANGES_REQUESTED** [review]
Put all of the message methods and GetExecutionType method at the bottom of the Function.

The testing methodology needs to change.
Add 1 to each field when you update it. Don't set every row to the same set of values.
Then check against the original list that it field value in each row has been incremented by 1.

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml (hunk 4 above)
Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: this duplicate line has been removed

File: Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml (hunk 5 above)
Comment (Netravati1205) [reply Partition (WBP)/Functions/VI0034C.VTI0034.fun.yml]: updated the correct ticket number.

**Comment by Netravati1205** [comment]
Testing methodology is modified based on the comment.

**Pull request merged**